File: sach/de3332.cpp
//DE3332.CPP - Tran Huu Nam - hsgtin.vn - 2024
#include <bits/stdc++.h>
using namespace std;
const int N=1e5+1;
int n,m,p=0;  int P[N]={}, C[N]={};
vector<int> ke[N];
void dfs(int u, int y){
  P[u]=y; C[y]++;
  for (auto v:ke[u])
    if (P[v]==0)   dfs(v, y);
}
int main(){
  ios_base::sync_with_stdio(false); cin.tie(NULL);
  cin>>n>>m;
  for (int i=1; i<=m; i++){
    int x,y; cin>>x>>y;
    ke[x].push_back(y);
    ke[y].push_back(x);
  }
  for (int i=1; i<=n; i++){
    if (P[i]==0){
      p++;
      dfs(i,p);
    }
  }
  int v=2;
  for (int i=3; i<=p; i++)
      if (C[i]>C[v]) v=i;
  cout<<C[1]+C[v];
  return 0;
}

File: sach/de3332.py
//DE3332.PY - Tran Huu Nam - hsgtin.vn - 2024
import sys
from collections import defaultdict
N = 100001
n, m, p = 0, 0, 0
P = [0] * N
C = [0] * N
ke = defaultdict(list)
def dfs(u, y):
    P[u] = y
    C[y] += 1
    for v in ke[u]:
        if P[v] == 0:
            dfs(v, y)
def main():
    input = sys.stdin.read
    data = input().split()
    index = 0
    n = int(data[index])
    index += 1
    m = int(data[index])
    index += 1
    for _ in range(m):
        x = int(data[index])
        index += 1
        y = int(data[index])
        index += 1
        ke[x].append(y)
        ke[y].append(x)
    for i in range(1, n + 1):
        if P[i] == 0:
            p += 1
            dfs(i, p)
    v = 2
    for i in range(3, p + 1):
        if C[i] > C[v]:
            v = i
    print(C[1] + C[v])
if __name__ == "__main__":
    main()

